deployment/docker-compose.yml

```diff
@@ -0,0 +1,37 @@
+# This file is meant for local development
+
+version: "3.9"
+
+services:
+  tf:
+    image: hashicorp/terraform:1.4.0
+    working_dir: /deployment
+    volumes:
+      - .:/deployment
+
+  scw:
+    image: scaleway/cli:v2.12.0
+
+  tf-docs:
+    image: quay.io/terraform-docs/terraform-docs:0.16.0
+    user: "${USER_UID:-1000}:0"
+    working_dir: /deployment
+    entrypoint: terraform-docs markdown --recursive --recursive-path ../../modules --output-file README.md
+    command: environments/staging
+    volumes:
+      - .:/deployment
+
+  tf-fmt:
+    image: hashicorp/terraform:1.4.0
+    working_dir: /deployment
+    command: fmt -recursive
+    volumes:
+      - .:/deployment
+
+  tf-vars:
+    image: quay.io/terraform-docs/terraform-docs:0.16.0
+    user: "${USER_UID:-1000}:0"
+    working_dir: /deployment
+    entrypoint: terraform-docs tfvars --output-file "terraform.tfvars.json" json
```

Comment:
@vmttn Je tombe sur un os avec cette commande :
```
$ docker compose run --rm tf-vars environments/staging

environments/staging/terraform.tfvars.json updated successfully

$ cat environments/staging/terraform.tfvars.json

<!-- BEGIN_TF_DOCS -->
{
…
}
<!-- END_TF_DOCS -->
```

Bien sûr les marqueurs `<!-- BEGIN_TF_DOCS -->` rendent le fichier JSON invalide.

Quand on enlève l'option `--output-file "terraform.tfvars.json"` de l'appel à `terraform-docs tfvars`, le JSON qui sort est correct (sans les commentaires), mais bien sûr il n'est pas écrit dans le fichier. Un bug de `terraform-docs` ?

Reply:
Je ne crois pas que ce soit un bug. `terraform-docs` ajoute une section au fichier `terraform.tfvars.json`, car celui peut préexister et contenir des valeurs que l'on souhaite conserver. Les balises permettent de répérer ce qui a été généré.

Reply:
Le fichier existe uniquement pour le développemment.

L'autre option (privilégiée pour la prod) : définir les variables via l'environnment : `TF_VAR_datawarehouse_admin_password=...` etc.

terraform ne gère par nativement les fichiers `.env` à l'instar de docker compose, mais il suffirait en théorie d'utiliser https://github.com/direnv/direnv, que je n'ai pas privilégié pour ne pas ajouter un prerequis au projet

Reply:
Je comprends l'utilisation des commentaires par `terraform-docs` pour identifier les parties à remplacer dans le fichier cible mais :
 - je constate que la partie entre `<!-- BEGIN_TF_DOCS -->` et `<!-- END_TF_DOCS -->` est entièrement remplacée lors de l'appel à `terraform-docs tfvars`, donc aucune valeur n'est conservée.
 - dans le cas d'un fichier JSON comme `terraform.tfvars.json`, la présence de commentaires comme `<!-- BEGIN_TF_DOCS -->`  rend le fichier simplement inutilisable car ce n'est plus du JSON valide.

Si j'ai remonté ce problème c'est parce que j'essayais simplement de suivre les instructions données dans le `README` :
> To generate a configuration file for a specific environment:

> 
> ```bash
> USER_ID=$(id -u) docker compose run --rm tf-vars environments/<ENVIRONMENT>
> ```
> 
> The generated `terraform.tfvars.json` file can be filled with the appropriate values for that environment.

Or, en suivant ces instructions, lors du lancement ensuite de `terraform plan` j'obtiens ces erreurs :

```
╷
│ Error: Invalid start of value
│
│   on terraform.tfvars.json line 1:
│    1: <!-- BEGIN_TF_DOCS -->
│
│ A JSON value must start with a brace, a bracket, a number, a string, or a
│ keyword.
╵
╷
│ Error: Root value must be object
│
│   on terraform.tfvars.json line 1:
│    1: <!-- BEGIN_TF_DOCS -->
│
│ The root value in a JSON-based configuration must be either a JSON object or
│ a JSON array of objects.
╵
```

En l'état, pour que ces instructions fonctionnent il faudrait ajouter "remove the first and last line of the generated `terraform.tfvars.json`", ou quelque chose comme ça.

Alternativement, j'ai constaté qu'en ajoutant les options `--output-mode replace --output-template '{{ .Content }}'` à l'invocation de `terraform-docs tfvars` on obtenait bien un fichier JSON sans les commentaires:

```suggestion
    entrypoint: terraform-docs tfvars --output-file "terraform.tfvars.json" --output-mode replace --output-template '{{ .Content }}' json
```

Reply:
Chez moi le bloc `<!-- BEGIN_TF_DOCS --> ...` est bien ajouté à la suite :man_shrugging: 

L'inconvénient de remplacer le fichier c'est qu'on va perdre des variables déjà définies.

Dans tous les cas, le fichier doit être au minimum édité pour y configurer les variables.

Est-ce qu'il ne suffit pas de préciser dans la documentation que le fichier doit être édité pour être un fichier json valide ? Ces étapes n'ont pas pour but d'être complétement automatisées, mais simplement d'initier le projet pour développer en local